staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/builder.go

```diff
@@ -322,6 +324,42 @@ func (b *builder) buildKubeNative(schema *structuralschema.Structural, v2 bool)
 	return ret
 }
 
+func addEmbeddedProperties(s *spec.Schema) {
+	if s == nil {
+		return
+	}
+	for k := range s.Properties {
+		v := s.Properties[k]
+		addEmbeddedProperties(&v)
+		s.Properties[k] = v
+	}
+	if s.Items != nil {
+		addEmbeddedProperties(s.Items.Schema)
+	}
+	if isTrue, ok := s.VendorExtensible.Extensions.GetBool("x-kubernetes-embedded-resource"); ok && isTrue {
+		addTypeMetaProperties(s)
```

Comment:
this overrides user-defined schema

Reply:
Yes, but just for publishing, right?

Reply:
I am fine with overriding for now. See my other comment about `$ref` and properties in kubectl. I think we have no choice other than throwing away in publishing what the user added and just publish the `$ref`.

Reply:
That's a reasonable starting point for 1.15. We can look into combining as part of publishing v3

Reply:
yeah I mentioned this because I thought the code was going in a different direction... glad I was wrong :)

> See my other comment about $ref and properties in kubectl

@sttts you're right. at least `kubectl create` [doesn't tolerate](https://github.com/kubernetes/kubernetes/blob/master/vendor/k8s.io/kube-openapi/pkg/util/proto/document.go#L96) schema with both `$ref` and properties

```
diff --git a/staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/builder.go b/staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/builder.go
index 6701995ddb..33c083610b 100644
--- a/staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/builder.go
+++ b/staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/builder.go
@@ -311,7 +311,8 @@ func (b *builder) buildKubeNative(schema *structuralschema.Structural, v2 bool)
                // TODO(roycaihw, 1.16): go through unfolding and v3->v2 conversion following the KEP, if the root schema
                // is a x-kubernetes-embedded-resource
                ret.SetProperty("metadata", *spec.RefSchema(objectMetaSchemaRef).
-                       WithDescription(swaggerPartialObjectMetadataDescriptions["metadata"]))
+                       WithDescription(swaggerPartialObjectMetadataDescriptions["metadata"]).
+                       SetProperty("testBool", *spec.BooleanProperty()))
```
```
$ curl localhost:8080/openapi/v2 | jq '.definitions["com.example.stable.v1.Bar"].properties["metadata"]'
{
  "description": "Standard object's metadata. More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata",
  "properties": {
    "testBool": {
      "type": "boolean"
    }
  },
  "$ref": "#/definitions/io.k8s.apimachinery.pkg.apis.meta.v1.ObjectMeta"
}
```
create some valid CR
```
$ cat bar.yaml 
apiVersion: stable.example.com/v1
kind: Bar
metadata:
  name: foov1
```
```
$ kubectl create -f bar.yaml 
error: SchemaError(com.example.stable.v1.Bar.metadata): unallowed embedded type definition
```
somehow `kubectl apply -f` wasn't doing client-side validation for me. I will track that later
```
$ kubectl apply -f bar.yaml 
bar.stable.example.com/foov1 created
$ kubectl apply -f bar.yaml --validate=true
bar.stable.example.com/foov1 unchanged
```